File: tests/sdks/python/containers/localstack_container.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import boto3
from testcontainers.localstack import (
    LocalStackContainer as BaseLocalStackContainer,
)

from envilder.application.envilder_client import EnvilderClient
from envilder.application.map_file_parser import MapFileParser
from envilder.domain.map_file_config import MapFileConfig
from envilder.domain.secret_provider_type import SecretProviderType
from envilder.infrastructure.aws.aws_ssm_secret_provider import (
    AwsSsmSecretProvider,
)
from envilder.infrastructure.secret_provider_factory import (
    SecretProviderFactory,
)

_SECRETS_MAP = Path(__file__).resolve().parent.parent / "secrets-map.json"


class LocalStackContainer:
    def __init__(self) -> None:
        self._container: BaseLocalStackContainer | None = None
        self._endpoint_url: str = ""

    def start(self) -> "LocalStackContainer":
        print("\n[LocalStack] Starting container...")

        environment = self._load_environment()
        if not environment.get("LOCALSTACK_AUTH_TOKEN"):
            raise EnvironmentError(
                "LOCALSTACK_AUTH_TOKEN could not be resolved" " from secrets-map.json"
            )

        self._container = BaseLocalStackContainer("localstack/localstack:stable")
        for key, value in environment.items():
            self._container.with_env(key, value)

        self._container.start()
        self._endpoint_url = self._container.get_url()

        print(f"[LocalStack] Ready at: {self._endpoint_url}")
        return self

    def stop(self) -> None:
        if self._container:
            print("[LocalStack] Stopping container...")
            self._container.stop()
            self._container = None

    def get_endpoint_url(self) -> str:
        return self._endpoint_url

    def get_ssm_client(self) -> Any:
        return boto3.client(
            "ssm",
            endpoint_url=self._endpoint_url,
            region_name="us-east-1",
            aws_access_key_id="test",
            aws_secret_access_key="test",
        )

    def create_provider(self) -> AwsSsmSecretProvider:
        return AwsSsmSecretProvider(self.get_ssm_client())

    @staticmethod
    def _load_environment() -> dict[str, str]:
        json_content = _SECRETS_MAP.read_text(encoding="utf-8")
        parser = MapFileParser()
        map_file = parser.parse(json_content)
        client = LocalStackContainer._resolve_client(map_file.config)
        return client.resolve_secrets(map_file)

    @staticmethod
    def _resolve_client(
        config: MapFileConfig,
    ) -> EnvilderClient:
        try:
            provider = SecretProviderFactory.create(config)
            return EnvilderClient(provider)
        except Exception:
            fallback = MapFileConfig(provider=SecretProviderType.AWS)
            return EnvilderClient(SecretProviderFactory.create(fallback))

File: tests/sdks/typescript/containers/localstack-container.ts
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { SSMClient } from '@aws-sdk/client-ssm';
import {
  LocalstackContainer,
  type StartedLocalStackContainer,
} from '@testcontainers/localstack';
import { EnvilderClient } from '../../../../src/sdks/typescript/src/application/envilder-client.js';
import { MapFileParser } from '../../../../src/sdks/typescript/src/application/map-file-parser.js';
import type { MapFileConfig } from '../../../../src/sdks/typescript/src/domain/map-file-config.js';
import { SecretProviderType } from '../../../../src/sdks/typescript/src/domain/secret-provider-type.js';
import { AwsSsmSecretProvider } from '../../../../src/sdks/typescript/src/infrastructure/aws/aws-ssm-secret-provider.js';
import { createSecretProvider } from '../../../../src/sdks/typescript/src/infrastructure/secret-provider-factory.js';

const LOCALSTACK_IMAGE = 'localstack/localstack:stable';
const SECRETS_MAP = path.resolve(__dirname, '../../../../secrets-map.json');

export class LocalStackTestContainer {
  private container: StartedLocalStackContainer | null = null;
  private endpointUrl = '';

  async start(): Promise<LocalStackTestContainer> {
    console.log('\n[LocalStack] Starting container...');

    const environment = await loadEnvironment();
    if (!environment.get('LOCALSTACK_AUTH_TOKEN')) {
      throw new Error(
        'LOCALSTACK_AUTH_TOKEN could not be resolved from secrets-map.json',
      );
    }

    const envRecord: Record<string, string> = {};
    for (const [key, value] of environment) {
      envRecord[key] = value;
    }

    this.container = await new LocalstackContainer(LOCALSTACK_IMAGE)
      .withName('localstack')
      .withEnvironment(envRecord)
      .start();

    this.endpointUrl = this.container.getConnectionUri();
    console.log(`[LocalStack] Ready at: ${this.endpointUrl}`);
    return this;
  }

  async stop(): Promise<void> {
    if (this.container) {
      console.log('[LocalStack] Stopping container...');
      await this.container.stop();
      this.container = null;
    }
  }

  getEndpointUrl(): string {
    return this.endpointUrl;
  }

  getSsmClient(): SSMClient {
    return new SSMClient({
      endpoint: this.endpointUrl,
      region: 'us-east-1',
      credentials: {
        accessKeyId: 'test',
        secretAccessKey: 'test',
      },
    });
  }

  createProvider(): AwsSsmSecretProvider {
    return new AwsSsmSecretProvider(this.getSsmClient());
  }
}

async function loadEnvironment(): Promise<Map<string, string>> {
  const json = await readFile(SECRETS_MAP, 'utf-8');
  const parser = new MapFileParser();
  const mapFile = parser.parse(json);
  const client = resolveClient(mapFile.config);
  return client.resolveSecrets(mapFile);
}

function resolveClient(config: MapFileConfig): EnvilderClient {
  try {
    const provider = createSecretProvider(config);
    return new EnvilderClient(provider);
  } catch {
    const fallback = createSecretProvider({
      provider: SecretProviderType.Aws,
    });
    return new EnvilderClient(fallback);
  }
}
